exporter/chronicleexporter/marshal.go

```diff
@@ -28,83 +28,71 @@ import (
 }
 
+type payload struct {
+	Entries    []entry `json:"entries"`
+	CustomerID string  `json:"customer_id"`
+	LogType    string  `json:"log_type"`
+}
+
+type entry struct {
+	LogText   string `json:"log_text"`
+	Timestamp string `json:"ts_rfc3339"`
+}
+
+type logMarshaler interface {
+	MarshalRawLogs(ctx context.Context, ld plog.Logs) ([]payload, error)
+}
+
 func newMarshaler(cfg Config, teleSettings component.TelemetrySettings) *marshaler {
 	return &marshaler{
 		cfg:          cfg,
 		teleSettings: teleSettings,
 	}
 }
 
-func (m *marshaler) MarshalRawLogs(ctx context.Context, ld plog.Logs) ([]byte, error) {
+func (m *marshaler) MarshalRawLogs(ctx context.Context, ld plog.Logs) ([]payload, error) {
 	rawLogs, err := m.extractRawLogs(ctx, ld)
 	if err != nil {
 		return nil, fmt.Errorf("extract raw logs: %w", err)
 	}
 
-	rawLogData := map[string]any{
-		"entries":  rawLogs,
-		"log_type": m.cfg.LogType,
-	}
-
-	if m.cfg.CustomerID != "" {
-		rawLogData["customer_id"] = m.cfg.CustomerID
-	}
-
-	return json.Marshal(rawLogData)
-}
-
-type entry struct {
-	LogText   string `json:"log_text"`
-	Timestamp string `json:"ts_rfc3339"`
+	return m.constructPayloads(rawLogs), nil
 }
 
-func (m *marshaler) extractRawLogs(ctx context.Context, ld plog.Logs) ([]entry, error) {
-	entries := []entry{}
+func (m *marshaler) extractRawLogs(ctx context.Context, ld plog.Logs) (map[string][]entry, error) {
+	entries := make(map[string][]entry)
 
 	for i := 0; i < ld.ResourceLogs().Len(); i++ {
 		resourceLog := ld.ResourceLogs().At(i)
 		for j := 0; j < resourceLog.ScopeLogs().Len(); j++ {
 			scopeLog := resourceLog.ScopeLogs().At(j)
 			for k := 0; k < scopeLog.LogRecords().Len(); k++ {
 				logRecord := scopeLog.LogRecords().At(k)
+				rawLog, logType, err := m.processLogRecord(ctx, logRecord, scopeLog.Scope(), resourceLog.Resource())
+				if err != nil {
+					m.teleSettings.Logger.Error("Error processing log record", zap.Error(err))
+					continue
+				}
 
-				var rawLog string
-				var err error
-				if m.cfg.RawLogField == "" {
-					body := logRecord.Body().Str()
-					entireLogRecord := map[string]any{
-						"body":                body,
-						"attributes":          logRecord.Attributes().AsRaw(),
-						"resource_attributes": resourceLog.Resource().Attributes().AsRaw(),
-					}
-
-					bytesLogRecord, err := json.Marshal(entireLogRecord)
-					if err != nil {
-						return nil, fmt.Errorf("marshal log record: %w", err)
-					}
-
-					rawLog = string(bytesLogRecord)
-				} else {
-					rawLog, err = m.getRawField(ctx, logRecord, scopeLog.Scope(), resourceLog.Resource())
-					if err != nil {
-						m.teleSettings.Logger.Error("Error getting raw field", zap.Error(err))
-						continue
-					}
+				if rawLog == "" {
+					continue
 				}
 
-				entries = append(entries, entry{
+				entries[logType] = append(entries[logType], entry{
 					LogText:   rawLog,
 					Timestamp: logRecord.Timestamp().AsTime().Format(time.RFC3339Nano),
```

Comment:
This is tangential to this PR, but we should fall back to ObservedTimestamp if the timestamp is unset (zero value)